### DISCUSSION TITLE
Let's Encrypt

### DISCUSSION THREAD

**Discussion by almereyda** [staticdeploy/staticdeploy#61] [Ideas]
This is a nice alternative to deploying static websites. However, I haven't found any documentation or note about how this integrates with Let's Encrypt, and allows me to deploy my sites to an HTTPS endpoint.

The instructions in https://staticdeploy.io/docs/getting-started-quickstart also only bind to port 80. Therefore I am a little confused how this is to be used in a real-world scenario. Can you help out with pointers?

**Comment by pscanf** [comment]
Hi @almereyda.

Currently StaticDeploy does not directly integrate with Let's Encrypt, or in general expose an HTTPS endpoint. This is more or less by design, as StaticDeploy's static server only focuses on **correctly serving files**, and leaves the job of **securing how files are served** to _someone else_. That someone else could be for instance an nginx reverse proxy sitting in front of StaticDeploy and configured to obtain certificates from Let's Encrypt

But I can definitely see how having StaticDeploy directly obtaining certificates could be useful. However for now all deployments I'm using StaticDeploy for are behind either a CDN or a reverse proxy providing SSL termination, so I haven't felt the need for this feature yet. I can add it to the backlog for future evolutions though.

**Comment by almereyda** [comment]
Hi @pscanf ,

thank you for your detailed answer. My intention derives from knowing https://static.land/ and how it manages Let's Encrypt all by itself. Unfortunately it requires its own dedicated IP, which is a precious resource in v4 land. My question here is actually more about informing an existing reverse proxy about domains that are configured and changed in and are served by StaticDeploy.

Given one would choose to declare Træfik a first-class citizen in StaticDeploy land, there is a chance one does not need to double the HTTP(S) endpoints of the reverse proxy by deploying StaticDeploy alongside. Theoretically one could allow it to communicate directly with the reverse proxy and update its configuration in real-time. This is possible with [Traefik Providers](https://docs.traefik.io/v2.0/providers/overview/#supported-providers).

- [Docker (Swarm) - Traefik](https://docs.traefik.io/v2.0/providers/docker/)
  Docker Engine allows to set container labels at runtime only, but Docker Swarm in particular knows an API command `docker service update` that has a `--container-label-add` and a `--container-label-rm` option. This means the domains for the StaticDeploy container reported to Træfik could be updated when the system changes. A prerequisite is to mount `docker.sock` into the container and implicitly allowing StaticDeploy root access to the Docker host, or by providing an auxiliary service that does so with reading the SD configuration only.
- [Kubernetes IngressRoute - Traefik](https://docs.traefik.io/v2.0/providers/kubernetes-crd/)
  Kubernetes [Labels and Selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) "can be attached to objects at creation time and subsequently added and modified at any time", too. This happens through the Kubernetes API, which StaticLand could learn to speak.
- [File - Traefik](https://docs.traefik.io/v2.0/providers/file/)
  The file provider consumes configuration files that are watched with `fsnotify` and updates the proxy configuration accordingly. In a Docker environment, such a configuration file could be generated by StaticDeploy and be host-volume mapped between the two containers. There might be similar facilities for other reverse proxies that I am currently not aware of.

From a point of view that takes security and ease of implementation into account, I wonder if it appears imaginable for you that there could be some kind of plugin/module that generates a Træfik configuration file for StaticDeploy, which is then always updated with the currently configured domains?

Additionally some kind of DNS checks could be put in place, similar to what GitHub does with GitHub Pages, in so Let's Encrypt certificates are created reliably, and the domain is well connected to the web server.

**Comment by pscanf** [comment]
> From a point of view that takes security and ease of implementation into account, I wonder if it appears imaginable for you that there could be some kind of plugin/module that generates a Træfik configuration file for StaticDeploy, which is then always updated with the currently configured domains?

The idea sounds good to me. I'd have to think about the details, but some sort of listener that watches `operationLogs` and generates/updates the Traefik configuration based on served domains sounds doable.

> Additionally some kind of DNS checks could be put in place, similar to what GitHub does with GitHub Pages, in so Let's Encrypt certificates are created reliably, and the domain is well connected to the web server.

This also sounds good. I already had in mind to add something like status checks that an admin user could see when visiting the management console. A DNS check could be a part of them, though maybe in the scenario you have in mind a failed check would be the trigger for some corrective action?

**Comment by almereyda** [comment]
To the last question: Indeed, if StaticDeploy connects to known DNS APIs.

There is https://github.com/AnalogJ/lexicon, a Python API, which connects to a plethora of them.

It appears important to highlight support for PowerDNS, since it is one of the few
DNS APIs that one can deploy on one's own.
